core/state/pruner/pruner.go

```diff
@@ -135,11 +135,14 @@ func prune(maindb ethdb.Database, stateBloom *stateBloom, start time.Time) error
 			if isCode {
 				checkKey = codeKey
 			}
-			// Filter out the state belongs the pruning target
-			if ok, err := stateBloom.Contain(checkKey); err != nil {
-				return err
-			} else if ok {
-				continue
+			if _, exist := middleStateRoots[common.BytesToHash(checkKey)]; exist {
+				log.Debug("Forcibly delete the middle state roots", "hash", common.BytesToHash(checkKey))
```

Comment:
This is more than a debug, IMO, should almost be a warning, since it really should not happen

Reply:
It happens a lot. Because they will be iterated for sure. It's not false-positive. Because we don't even check the bloom filter for them.

Reply:
Hey wait!
What if this is a clique chain, and all the last N (say, 300) blocks have the exact identical stateroot?

Reply:
Ah, I think that's fine, they won't be added into the `middleStateRoots`... _I think_